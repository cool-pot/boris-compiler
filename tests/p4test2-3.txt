*** test level 2, foreach array
global myarray;
array myarray[10..12];
myarray[10] = 0;
myarray[11] = 1;
myarray[12] = 2;

foreach myitem in myarray do
    print myitem;
end for
