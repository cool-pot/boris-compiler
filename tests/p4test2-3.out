; ModuleID = 'module-boris-input'
source_filename = "module-boris-input"
target triple = "x86_64-pc-linux-gnu"

@format_int = private unnamed_addr constant [4 x i8] c"%d\0A\00"

define i32 @strlen(i8* %s) {
init:
  %i = alloca i32
  store i32 0, i32* %i
  br label %check

check:                                            ; preds = %body, %init
  %0 = load i32, i32* %i
  %1 = getelementptr i8, i8* %s, i32 %0
  %2 = load i8, i8* %1
  %3 = icmp ne i8 0, %2
  br i1 %3, label %body, label %end

body:                                             ; preds = %check
  %4 = load i32, i32* %i
  %5 = add i32 %4, 1
  store i32 %5, i32* %i
  br label %check

end:                                              ; preds = %check
  %6 = load i32, i32* %i
  ret i32 %6
}

declare i32 @printf(...)

define void @main() {
MainEntry:
  %myarray = alloca i32, i32 3
  %0 = getelementptr i32, i32* %myarray, i32 0
  store i32 0, i32* %0
  %1 = getelementptr i32, i32* %myarray, i32 1
  store i32 1, i32* %1
  %2 = getelementptr i32, i32* %myarray, i32 2
  store i32 2, i32* %2
  br label %foreacha_init_

foreacha_init_:                                   ; preds = %MainEntry
  %_temp_id_0 = alloca i32
  %myitem = alloca i32
  store i32 0, i32* %_temp_id_0
  %load_singleid_int_temp = load i32, i32* %_temp_id_0
  %3 = getelementptr i32, i32* %myarray, i32 %load_singleid_int_temp
  %4 = load i32, i32* %3
  store i32 %4, i32* %myitem
  br label %foreacha_check_

foreacha_check_:                                  ; preds = %foreacha_body_, %foreacha_init_
  %load_singleid_int_temp1 = load i32, i32* %_temp_id_0
  %5 = icmp slt i32 %load_singleid_int_temp1, 3
  br i1 %5, label %foreacha_body_, label %foreacha_end_

foreacha_body_:                                   ; preds = %foreacha_check_
  %load_singleid_int_temp2 = load i32, i32* %myitem
  %6 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int, i32 0, i32 0), i32 %load_singleid_int_temp2)
  %load_singleid_int_temp3 = load i32, i32* %_temp_id_0
  %7 = add i32 %load_singleid_int_temp3, 1
  store i32 %7, i32* %_temp_id_0
  %load_singleid_int_temp4 = load i32, i32* %_temp_id_0
  %8 = getelementptr i32, i32* %myarray, i32 %load_singleid_int_temp4
  %9 = load i32, i32* %8
  store i32 %9, i32* %myitem
  br label %foreacha_check_

foreacha_end_:                                    ; preds = %foreacha_check_
  br label %EndEntry

EndEntry:                                         ; preds = %foreacha_end_
  ret void
}
